File: art_e9i8/circles.py
#!/usr/bin/env python3
"""MO 514722: exhaustive census of circles through >=3 lattice points with
circumradius <= R_MAX.  Exact integer arithmetic throughout.

Every circle through >=3 lattice points, translated so one of its lattice
points is the origin, satisfies  A(x^2+y^2) + Gx + Fy = 0  with integers
A>0, G, F, gcd(A,G,F)=1.  Center (-G/2A, -F/2A), r^2 = (G^2+F^2)/4A^2.
Two anchored circles are plane-translates iff (A, -G mod 2A, -F mod 2A,
r^2) agree, since re-anchoring at another lattice point shifts the center
by integers.

Rigor for the minimum-interior question (exactly 5 on-points):
  closed-disk lattice count >= ceil(pi (r - 1/sqrt2)^2)   [centered unit
  squares covering the shrunk disk], so interior >= pi(r-0.7072)^2 - 5.
  interior <= 105 forces r <= 6.63.  We search r <= R_MAX >= 7.3, which
  also covers the coarser bound pi(r-sqrt2)^2 <= 106 (r <= 7.23).
"""
import sys, math, json
from math import gcd, isqrt
import numpy as np
from collections import defaultdict

R_MAX = float(sys.argv[1]) if len(sys.argv) > 1 else 7.3
OUT = sys.argv[2] if len(sys.argv) > 2 else f"census_r{R_MAX:g}.json"
R2CAP_NUM = None  # r^2 <= R_MAX^2 exact via cross-multiplied ints below
D = 2 * R_MAX     # chord bound
DI = int(math.floor(D))

pts = [(x, y) for x in range(-DI, DI + 1) for y in range(-DI, DI + 1)
       if x * x + y * y <= D * D and (x, y) != (0, 0)]
pts.sort()
n = len(pts)
print(f"R_MAX={R_MAX}  candidate points={n}  pairs~{n*(n-1)//2}")

R2q = R_MAX * R_MAX
seen = set()
circles = []          # (A,G,F) anchored, normalized
for i in range(n):
    bx, by = pts[i]
    nb = bx * bx + by * by
    for j in range(i + 1, n):
        cx, cy = pts[j]
        dx, dy = cx - bx, cy - by
        if dx * dx + dy * dy > D * D:
            continue
        det = bx * cy - by * cx
        if det == 0:
            continue
        nc = cx * cx + cy * cy
        A = det
        G = -nb * cy + nc * by
        F = nb * cx - nc * bx
        if A < 0:
            A, G, F = -A, -G, -F
        g = gcd(gcd(A, abs(G)), abs(F))
        A //= g; G //= g; F //= g
        # r^2 = (G^2+F^2)/(4A^2) <= R2q ?
        num = G * G + F * F
        if num > R2q * 4 * A * A:
            continue
        key = (A, (-G) % (2 * A), (-F) % (2 * A), num)
        # num/(4A^2) with gcd(A,G,F)=1 is already a translation invariant
        # paired with A; (num, A) determines r^2.
        if key in seen:
            continue
        seen.add(key)
        circles.append((A, G, F))
print(f"unique circles (translation classes): {len(circles)}")

# exact counting per circle
by_k = defaultdict(list)   # on-count -> list of (interior, num, A, G, F)
for (A, G, F) in circles:
    cxf = -G / (2 * A); cyf = -F / (2 * A)
    r = math.sqrt(G * G + F * F) / (2 * A)
    x0, x1 = math.ceil(cxf - r - 1e-9), math.floor(cxf + r + 1e-9)
    y0, y1 = math.ceil(cyf - r - 1e-9), math.floor(cyf + r + 1e-9)
    xs = np.arange(x0, x1 + 1, dtype=np.int64)
    ys = np.arange(y0, y1 + 1, dtype=np.int64)
    X, Y = np.meshgrid(xs, ys, indexing="ij")
    val = A * (X * X + Y * Y) + G * X + F * Y
    on = int((val == 0).sum())
    inside = int((val < 0).sum())
    by_k[on].append((inside, G * G + F * F, A, G, F))

summary = {}
for k in sorted(by_k):
    lst = sorted(by_k[k])
    best = lst[0]
    # min radius representative too
    byr = min(lst, key=lambda t: t[1] / (4 * t[2] * t[2]))
    summary[k] = dict(
        count=len(lst),
        min_interior=best[0],
        min_interior_circle=dict(A=best[2], G=best[3], F=best[4],
                                 r=math.sqrt(best[1]) / (2 * best[2])),
        min_radius=math.sqrt(byr[1]) / (2 * byr[2]),
        min_radius_interior=byr[0],
        min_radius_circle=dict(A=byr[2], G=byr[3], F=byr[4]),
    )
    print(f"k={k:3d}  circles={len(lst):7d}  min_interior={best[0]:5d} "
          f"(r={math.sqrt(best[1])/(2*best[2]):.4f})  "
          f"min_radius={summary[k]['min_radius']:.4f}")

# dump full census for the art (k, r2num, A, interior) per circle
art = {str(k): [[t[0], t[1], t[2], t[3], t[4]] for t in sorted(by_k[k])]
       for k in by_k}
with open(OUT, "w") as f:
    json.dump(dict(R_MAX=R_MAX, summary=summary, census=art), f)
print("wrote", OUT)

File: art_e9i8/circles.c
/* MO 514722 census engine: all circles through >=3 lattice points with
   circumradius <= RMAX, exact integer arithmetic, translation-dedup.
   Prints one line per unique circle with on-count >= 5:
       C k A G F num interior      (r^2 = num/(4A^2))
   and aggregates k=3,4 into radius-binned histograms:
       H k bin count min_interior   (bin = floor(r / 0.125))
   gcc -O3 -march=native circles.c -o circles -lm ; ./circles RMAX      */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <math.h>

static int64_t g64(int64_t a, int64_t b){ while(b){int64_t t=a%b;a=b;b=t;} return a<0?-a:a; }

/* open-addressing hash of 128-bit keys */
static uint64_t *H1, *H2; static size_t HN;      /* HN power of two */
static inline int insert(uint64_t k1, uint64_t k2){
    uint64_t h = k1*0x9E3779B97F4A7C15ULL ^ k2*0xC2B2AE3D27D4EB4FULL;
    h ^= h>>29; h *= 0xBF58476D1CE4E5B9ULL; h ^= h>>32;
    size_t i = h & (HN-1);
    for(;;){
        if(!H1[i] && !H2[i]){ H1[i]=k1; H2[i]=k2; return 1; }
        if(H1[i]==k1 && H2[i]==k2) return 0;
        i = (i+1) & (HN-1);
    }
}

typedef struct { int64_t A,G,F,num; } Circ;

int main(int argc, char**argv){
    double RMAX = argc>1 ? atof(argv[1]) : 32.0;
    double D = 2*RMAX; int DI = (int)floor(D);
    /* candidate points (excluding origin) within D of origin */
    int np=0, cap=(2*DI+1)*(2*DI+1);
    int *px = malloc(cap*4), *py = malloc(cap*4);
    for(int x=-DI;x<=DI;x++) for(int y=-DI;y<=DI;y++)
        if((x||y) && (double)x*x+(double)y*y <= D*D){ px[np]=x; py[np]=y; np++; }
    fprintf(stderr,"RMAX=%g points=%d\n", RMAX, np);

    HN = 1; while(HN < 1ULL<<26) HN<<=1;        /* 67M slots = 1GB */
    HN = 1ULL<<26;
    H1 = calloc(HN,8); H2 = calloc(HN,8);
    if(!H1||!H2){fprintf(stderr,"alloc fail\n");return 1;}

    size_t ncirc=0, ccap=1<<20;
    Circ *cs = malloc(ccap*sizeof(Circ));
    double R2q = RMAX*RMAX;
    for(int i=0;i<np;i++){
        int64_t bx=px[i], by=py[i], nb=bx*bx+by*by;
        for(int j=i+1;j<np;j++){
            int64_t cx=px[j], cy=py[j];
            int64_t dx=cx-bx, dy=cy-by;
            if((double)dx*dx+(double)dy*dy > D*D) continue;
            int64_t A = bx*cy - by*cx;
            if(!A) continue;
            int64_t nc = cx*cx+cy*cy;
            int64_t G = -nb*cy + nc*by;
            int64_t F =  nb*cx - nc*bx;
            if(A<0){A=-A;G=-G;F=-F;}
            int64_t g = g64(g64(A,G),F);
            A/=g; G/=g; F/=g;
            int64_t num = G*G + F*F;
            if((double)num > R2q*4.0*A*A) continue;
            int64_t twoA = 2*A;
            int64_t gm = (-G) % twoA; if(gm<0) gm += twoA;
            int64_t fm = (-F) % twoA; if(fm<0) fm += twoA;
            /* pack: A<2^14, gm,fm<2^15, num<2^44 */
            uint64_t k1 = ((uint64_t)A<<30) | ((uint64_t)gm<<15) | (uint64_t)fm;
            uint64_t k2 = (uint64_t)num;
            if(insert(k1,k2)){
                if(ncirc==ccap){ ccap<<=1; cs=realloc(cs,ccap*sizeof(Circ)); }
                cs[ncirc].A=A; cs[ncirc].G=G; cs[ncirc].F=F; cs[ncirc].num=num;
                ncirc++;
            }
        }
    }
    free(H1); free(H2);
    fprintf(stderr,"unique circles: %zu\n", ncirc);

    enum{NB=400};
    long hist3[NB]; long hist4[NB]; long mi3[NB], mi4[NB];
    for(int b=0;b<NB;b++){hist3[b]=hist4[b]=0; mi3[b]=mi4[b]=-1;}

    for(size_t t=0;t<ncirc;t++){
        int64_t A=cs[t].A, G=cs[t].G, F=cs[t].F;
        double cxf = -G/(2.0*A), cyf = -F/(2.0*A);
        double r = sqrt((double)cs[t].num)/(2.0*A);
        int x0=(int)ceil(cxf-r-1e-9), x1=(int)floor(cxf+r+1e-9);
        int y0=(int)ceil(cyf-r-1e-9), y1=(int)floor(cyf+r+1e-9);
        int on=0; long in=0;
        for(int x=x0;x<=x1;x++){
            int64_t xx = A*(int64_t)x*x + G*x;
            for(int y=y0;y<=y1;y++){
                int64_t v = xx + A*(int64_t)y*y + F*y;
                if(v<0) in++; else if(v==0) on++;
            }
        }
        if(on>=5){
            printf("C %d %lld %lld %lld %lld %ld\n", on,(long long)A,
                (long long)G,(long long)F,(long long)cs[t].num, in);
        } else {
            int b=(int)(r/0.125); if(b>=NB)b=NB-1;
            if(on==3){ hist3[b]++; if(mi3[b]<0||in<mi3[b])mi3[b]=in; }
            else if(on==4){ hist4[b]++; if(mi4[b]<0||in<mi4[b])mi4[b]=in; }
        }
    }
    for(int b=0;b<NB;b++){
        if(hist3[b]) printf("H 3 %d %ld %ld\n",b,hist3[b],mi3[b]);
        if(hist4[b]) printf("H 4 %d %ld %ld\n",b,hist4[b],mi4[b]);
    }
    return 0;
}
